### PR TITLE
 [TAPE] Adding a Torch interface to the quantum tape

### PR DESCRIPTION
**Pull request by josh146**
**Context:** Adds a Torch interface to the quantum tape.

**Description of the Change:**

The quantum tape can now be interfaced with Torch as follows:
```python
dev = qml.device("default.qubit", wires=1)
p = torch.tensor([0.1, 0.2, 0.3], requires_grad=True)

with TorchInterface.apply(QuantumTape()) as qtape:
    qml.Rot(p[0], p[1] ** 2 + p[0] * p[2], p[1] * torch.sin(p[2]), wires=0)
    expval(qml.PauliX(0))

>>> result = qtape.execute(dev)
>>> print(result)
tensor([0.0698], dtype=torch.float64, grad_fn=<_TorchInterfaceBackward>)
>>> result.backward()
>>> print(p.grad)
tensor([0.2987, 0.3971, 0.0988])
```
In addition, you can construct the tape in advanced if you would like:

```python
with TorchInterface.apply(QuantumTape()) as qtape:
    qml.Rot(0, 0, 0, wires=0)
    expval(qml.PauliX(0))

p = torch.tensor([0.1, 0.2, 0.3], requires_grad=True)
weights = [p[0], p[1] ** 2 + p[0] * p[2], p[1] * torch.sin(p[2])]

qtape.set_parameters(weights, trainable_only=False)
qtape._update_trainable_params()
result = qtape.execute(dev)

>>> print(result)
tensor([0.0698], dtype=torch.float64, grad_fn=<_TorchInterfaceBackward>)
>>> result.backward()
>>> print(p.grad)
tensor([0.2987, 0.3971, 0.0988])
```

**Benefits:**
* Unlike the existing Torch interface, the new interface does not 'wrap' the QNode. So all QNode properties and attributes are still accessible.

**Possible Drawbacks:** n/a

**Related GitHub Issues:** n/a

### CODE REVIEW

File: tests/beta/interfaces/test_tape_torch.py [hunk 1]
```diff
@@ -0,0 +1,447 @@
...
+            expval(qml.PauliX(0))
+
+        assert tape.interface == "torch"
+        assert isinstance(tape, TorchInterface)
+        assert tape.__bare__ == QuantumTape
+
+        TorchInterface.apply(QuantumTape())
+        assert tape.interface == "torch"
+        assert isinstance(tape, TorchInterface)
+        assert tape.__bare__ == QuantumTape
```
Comment (co9olguy) [inline tests/beta/interfaces/test_tape_torch.py:53]: Not sure that `tape` has been reassigned at all

File: tests/beta/interfaces/test_tape_torch.py [hunk 2]
```diff
@@ -0,0 +1,447 @@
...
+        )
+
+        res = tape.execute(dev)
+        res.backward()
+
+        assert isinstance(params.grad, torch.Tensor)
+        assert params.shape == (2,)
+
+    def test_no_trainable_parameters(self, tol):
+        """Test evaluation and Jacobian if there are no trainable parameters"""
```
Comment (co9olguy) [inline tests/beta/interfaces/test_tape_torch.py:245]: jacobian actually (attempted to be) evaluated in this one :)

File: tests/beta/interfaces/test_tape_torch.py [hunk 3]
```diff
@@ -0,0 +1,447 @@
...
+                    np.cos(p_val[0]) * np.cos(p_val[1]) * np.cos(p_val[2])
+                    - np.sin(p_val[1]) * np.sin(p_val[2])
+                ),
+            ]
+        )
+        assert np.allclose(p.grad, expected, atol=tol, rtol=0)
+
+    def test_probability_differentiation(self, tol):
+        """Tests correct output shape and evaluation for a tape
+        with prob and expval outputs"""
```
Comment (co9olguy) [inline tests/beta/interfaces/test_tape_torch.py:355]: circuit does not have expval

File: tests/beta/interfaces/test_tape_torch.py [hunk 4]
```diff
@@ -0,0 +1,447 @@
...
+        with TorchInterface.apply(QuantumTape()) as tape:
+            qml.Hadamard(wires=[0])
+            qml.CNOT(wires=[0, 1])
+            sample(qml.PauliZ(0))
+            sample(qml.PauliX(1))
+
+        res = tape.execute(dev)
+
+        assert res.shape == (2, 10)
+        assert isinstance(res, torch.Tensor)
```
Comment (co9olguy) [inline tests/beta/interfaces/test_tape_torch.py:447]: No PassThru mode to check :wink:

File: pennylane/beta/interfaces/torch.py [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+        """Implements the forward pass QNode evaluation"""
+        # detach all input tensors, convert to NumPy array
+        ctx.args = args_to_numpy(input_)
+        ctx.kwargs = input_kwargs
+        ctx.save_for_backward(*input_)
+
+        tape = ctx.kwargs["tape"]
+        device = ctx.kwargs["device"]
+
+        # # unwrap constant parameters
```
Comment (co9olguy) [inline pennylane/beta/interfaces/torch.py:38]:
```suggestion
        # unwrap constant parameters

```

File: pennylane/beta/interfaces/torch.py [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+        tape = ctx.kwargs["tape"]
+        device = ctx.kwargs["device"]
+
+        tape.set_parameters(ctx.all_params_unwrapped, trainable_only=False)
+        jacobian = tape.jacobian(device, params=ctx.args, **tape.jacobian_options)
+        tape.set_parameters(ctx.all_params, trainable_only=False)
+
+        jacobian = torch.as_tensor(jacobian, dtype=grad_output.dtype)
+
+        vjp = torch.transpose(grad_output.view(-1, 1), 0, 1) @ jacobian
```
Comment (co9olguy) [inline pennylane/beta/interfaces/torch.py:73]: Think you can save a transpose here by using `view(1,-1)`

File: pennylane/beta/interfaces/torch.py [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+
+
+class TorchInterface(AnnotatedQueue):
+    """Mixin class for applying an Torch interface to a :class:`~.QuantumTape`.
+
+    Torch-compatible quantum tape classes can be created via subclassing:
+
+    .. code-block:: python
+
+        class MyAutogradQuantumTape(AutogradInterface, QuantumTape):
```
Comment (co9olguy) [inline pennylane/beta/interfaces/torch.py:95]:
```suggestion
        class MyTorchQuantumTape(TorchInterface, QuantumTape):

```

File: pennylane/beta/interfaces/torch.py [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+
+    Alternatively, the autograd interface can be dynamically applied to existing
```
Comment (co9olguy) [inline pennylane/beta/interfaces/torch.py:97]:
```suggestion
    Alternatively, the Torch interface can be dynamically applied to existing

```

File: pennylane/beta/interfaces/torch.py [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+    quantum tapes via the :meth:`~.apply` class method. This modifies the
+    tape **in place**.
+
+    Once created, the autograd interface can be used to perform quantum-classical
```
Comment (co9olguy) [inline pennylane/beta/interfaces/torch.py:101]:
```suggestion
    Once created, the Torch interface can be used to perform quantum-classical

```

File: pennylane/beta/interfaces/torch.py [hunk 10]
```diff
@@ -0,0 +1,189 @@
...
+    differentiable programming.
+
+    **Example**
+
+    One a Torch quantum tape has been created, it can be differentiated using the gradient tape:
```
Comment (co9olguy) [inline pennylane/beta/interfaces/torch.py:106]:
```suggestion
    Once a Torch quantum tape has been created, it can be differentiated using the gradient tape:

```

Comment (co9olguy) [reply pennylane/beta/interfaces/torch.py:106]: better to us torch terminology ("backward()", not "gradient tape")

File: tests/beta/interfaces/test_tape_torch.py [hunk 11]
```diff
@@ -0,0 +1,447 @@
...
+import numpy as np
+
+import pennylane as qml
+from pennylane.beta.tapes import QuantumTape
+from pennylane.beta.queuing import expval, var, sample, probs
+from pennylane.beta.interfaces.torch import TorchInterface
+
+
+class TestTorchQuantumTape:
+    """Test the autograd interface applied to a tape"""
```
Comment (co9olguy) [inline tests/beta/interfaces/test_tape_torch.py:28]:
```suggestion
    """Test the Torch interface applied to a tape"""

```

**Review by co9olguy — APPROVED** [review]

File: pennylane/beta/interfaces/torch.py (hunk 6 above)
Comment (josh146) [reply pennylane/beta/interfaces/torch.py:73]: Nice! Let me test it

File: tests/beta/interfaces/test_tape_torch.py (hunk 1 above)
Comment (josh146) [reply tests/beta/interfaces/test_tape_torch.py:53]: This is because `TorchInterface.apply` acts in-place, modifying the tape argument it is given.

I originally tried to implement a non-in-place method, of the form:

```python
new_tape = TorchInterface.apply(original_tape)
```
(which I think is more functional).

_But_, this requires doing a deep copy of the original tape, which works fine in autograd and TF, but not pytorch (torch tensors that require gradients do not support copying). So I changed this method to be in-place across the board.

Which, in hindsight, might be better/more efficient, developers just have to be aware of the side effect.

File: tests/beta/interfaces/test_tape_torch.py (hunk 3 above)
Comment (josh146) [reply tests/beta/interfaces/test_tape_torch.py:355]:
```suggestion
        with multiple prob outputs"""
```

File: tests/beta/interfaces/test_tape_torch.py (hunk 4 above)
Comment (josh146) [reply tests/beta/interfaces/test_tape_torch.py:447]: 😢

File: tests/beta/interfaces/test_tape_torch.py (hunk 1 above)
Comment (co9olguy) [reply tests/beta/interfaces/test_tape_torch.py:54]: what about `XYZInterface.replace(new_tape)` to be more explicit?

Comment (co9olguy) [reply tests/beta/interfaces/test_tape_torch.py:54]: or maybe it's just that you didn't bother writing any ops on the new tape that confused me

Comment (josh146) [reply tests/beta/interfaces/test_tape_torch.py:54]: Oops, no you're right, this is my mistake. The test has a typo, it should be
```python
        assert tape.interface == "torch"
        assert isinstance(tape, TorchInterface)
        assert tape.__bare__ == QuantumTape

        TorchInterface.apply(tape)
        assert tape.interface == "torch"
        assert isinstance(tape, TorchInterface)
        assert tape.__bare__ == QuantumTape
```

I'll fix this, and add an extra assertion just to make sure it _is_ the same tape in both cases.
